x/concentrated-liquidity/model/pool.go

```diff
@@ -193,36 +193,62 @@ func (p *Pool) UpdateLiquidityIfActivePosition(ctx sdk.Context, lowerTick, upper
 // -The position is active ( lowerTick <= p.CurrentTick < upperTick).
 //   - The provided liquidity is distributed in both tokens.
 //   - Actual amounts might differ from desired because we recalculate them from liquidity delta and sqrt price.
-//     the calculations lead to amounts being off. // TODO: confirm logic is correct
 //
 // - Current tick is below the position ( p.CurrentTick < lowerTick).
 //   - The provided liquidity is distributed in token0 only.
 //
 // - Current tick is above the position ( p.CurrentTick >= p.upperTick ).
 //   - The provided liquidity is distributed in token1 only.
 //
-// TODO: add tests.
-func (p Pool) CalcActualAmounts(ctx sdk.Context, lowerTick, upperTick int64, sqrtRatioLowerTick, sqrtRatioUpperTick sdk.Dec, liquidityDelta sdk.Dec) (actualAmountDenom0 sdk.Dec, actualAmountDenom1 sdk.Dec) {
+// Note, that liquidityDelta can be positive or negative but cannot be zero.
+// If zero, an error is returned.
+// If positive, we assume, liquidity being added. As a result, we round up so that
+// we request a user to add more liquidity in favor of the pool.
+// If negative, we assume, liquidity being removed. As a result, we round down so that
+// we request a user to remove less liquidity in favor of the pool.
+func (p Pool) CalcActualAmounts(ctx sdk.Context, lowerTick, upperTick int64, liquidityDelta sdk.Dec) (actualAmountDenom0 sdk.Dec, actualAmountDenom1 sdk.Dec, err error) {
+	if liquidityDelta.IsZero() {
+		return sdk.Dec{}, sdk.Dec{}, types.ErrZeroLiquidity
+	}
+
+	if lowerTick >= upperTick {
+		return sdk.Dec{}, sdk.Dec{}, types.InvalidLowerUpperTickError{LowerTick: lowerTick, UpperTick: upperTick}
+	}
+
+	// Transform the provided ticks into their corresponding sqrtPrices.
+	sqrtPriceLowerTick, sqrtPriceUpperTick, err := math.TicksToSqrtPrice(lowerTick, upperTick)
+	if err != nil {
+		return sdk.Dec{}, sdk.Dec{}, err
+	}
+
+	// When liquidity delta is positive, that means that we are adding liquidity.
+	// Therefore, we should round up to require user provide a higher amount
+	// in favor of the pool.
+	// When liquidity delta is negative, that means that we are removing liquidity.
+	// Therefore, we should round down to require user provide a lower amount
+	// in favor of the pool.
+	roundUp := liquidityDelta.IsPositive()
```

Comment:
oh wow, so it was just this error huh. Did you figure it out from this code? https://github.com/Jeiwan/uniswapv3-code/blob/main/src/lib/Math.sol#L61